File: fsharp/isbn-verifier/IsbnVerifier.fs
module IsbnVerifier

open System.Text.RegularExpressions

let parse (isbn: string) =
    match isbn.Replace("-", "") with
    | str when Regex.IsMatch(str, @"^\d{9}[\dX]$") -> Ok str
    | _ -> Error "invalid format"

let checksum isbn =
    let value c =
        match c with
        | 'X' -> 10
        | d -> System.Char.GetNumericValue(d) |> int

    isbn
    |> Seq.rev
    |> Seq.mapi (fun i c -> (value c) * (i + 1))
    |> Seq.sum
    |> fun x -> x % 11

let isValid isbn =
    match parse isbn with
    | Ok valid -> checksum valid = 0
    | Error _ -> false

File: csharp/isbn-verifier/IsbnVerifier.cs
using System.Linq;
using System.Text.RegularExpressions;

public static class IsbnVerifier
{
    public static bool IsValid(string number)
    {
        var digits = number.Replace("-", "");
        return IsValidFormat(digits) && IsValidChecksum(digits);
    }

    private static bool IsValidFormat(string digits) =>
        Regex.IsMatch(digits, @"^\d{9}[\dX]$");

    private static bool IsValidChecksum(string digits)
    {
        static int DigitValue(char digit) => digit == 'X' ? 10 : (int)char.GetNumericValue(digit);

        var sum = digits
            .Select(DigitValue)
            .Reverse()
            .Select((n, i) => n * (i + 1))
            .Sum();

        return sum % 11 == 0;
    }
}